vendor/github.com/openshift/library-go/pkg/image/registryclient/alternative-sources.go

```diff
@@ -0,0 +1,135 @@
+package registryclient
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+
+	"github.com/docker/distribution"
+	operatorv1alpha1 "github.com/openshift/api/operator/v1alpha1"
+	operatorv1alpha1scheme "github.com/openshift/client-go/operator/clientset/versioned/scheme"
+	operatorv1alpha1client "github.com/openshift/client-go/operator/clientset/versioned/typed/operator/v1alpha1"
+	imagereference "github.com/openshift/library-go/pkg/image/reference"
+)
+
+// AlternativeImageSources are ImageContentSourcePolicy variables to look up image sources
+type AlternativeImageSources struct {
+	ICSPFile     string
+	ICSPList     []operatorv1alpha1.ImageContentSourcePolicy
+	ICSPClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)
+}
+
+type AlternativeSources interface {
+	PreferredImageSource(imageRef imagereference.DockerImageReference) (imagereference.DockerImageReference, distribution.Repository, distribution.ManifestService, error)
+}
+
+func (c *Context) WithICSP(icspFile string, icspList []operatorv1alpha1.ImageContentSourcePolicy, icspClientFn func() (operatorv1alpha1client.ImageContentSourcePolicyInterface, error)) *Context {
```

Comment:
Why you need both the list and the client? I'd expect either not both.

Reply:
This still holds, I don't think you need both the actual list of ICSP and client, either of the two should be sufficient, no?

Reply:
From looking at the code it seems that `icspList` is not being used here nor in `pkg/cli/image/manifest/security.go` I'd just drop it. Also you can probably just pass file and client. The presence of either is a sign that a user requested ICSP. `pkg/cli/image/manifest/security.go` will use the flag only to read user intent, iow. user will pass `--lookup-cluster-icsp` or `--icsp-file` and you pass both to this method.

Reply:
don't need to pass the icspList,but the LookupClusterICSP is needed afaict. It has to be carried thru to differentiate from implicit icsp lookup from explicit lookup. I added comments to the library-go file to help clarify - I'll keep testing though to be sure.

Reply:
nm, updated & merged LookupClusterICSP w/ TryAlternativeSources.